admin/server/device_code.go

```diff
@@ -0,0 +1,315 @@
+package server
+
+import (
+	"crypto/rand"
+	"database/sql"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/gorilla/sessions"
+	"github.com/rilldata/rill/admin/database"
+	"github.com/rilldata/rill/cli/cmd/auth"
+	"go.uber.org/zap"
+)
+
+const (
+	tokenExpirationSeconds        = 60 * 10 // 10 minutes
+	deviceCodeGrantType           = "urn:ietf:params:oauth:grant-type:device_code"
+	DefaultDeviceVerificationHost = "https://ui.rilldata.com"
+)
+
+// DeviceCodeResponse encapsulates the response for obtaining a device code.
+type DeviceCodeResponse struct {
+	DeviceCode              string `json:"device_code"`
+	UserCode                string `json:"user_code"`
+	VerificationURI         string `json:"verification_uri"`
+	VerificationCompleteURI string `json:"verification_uri_complete"`
+	ExpiresIn               int    `json:"expires_in"`
+	PollingInterval         int    `json:"interval"`
+}
+
+// TokenRequest encapsulates the request for obtaining an access token.
+type TokenRequest struct {
+	GrantType  string `json:"grant_type"`
+	DeviceCode string `json:"device_code"`
+	ClientID   string `json:"client_id"`
+}
+
+// handleDeviceCodeRequest creates a 24 digit random device code and 8 digit user code and returns that
+// to the client. The device code is used to poll for an access token, and the user code is displayed
+// to the user and is used to authorize the device. The device code and user code are stored in the
+// server's device code store.
+func (s *Server) handleDeviceCodeRequest(w http.ResponseWriter, r *http.Request, pathParams map[string]string) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		s.logger.Error("failed to read request body", zap.Error(err))
+		internalServerError(w, err)
```

Comment:
1. it was not getting logged 
2. the log is for debugging at server side, internal server error is for user since we don't want to pass actual err to user.

Reply:
It should just return the error as an internal server error. Then a middleware can remove the detailed error info (if we don't want to show to user) and log it instead. 

Errors should generally not be logged in places where errors are returned. It's better to decide on logging at the boundaries (e.g. in middleware), otherwise you start adding error logging statements at every `if err != nil {...`.

Will add an issue to add logging middleware for these endpoints instead.

Reply:
Actually the implementation of internalServerError should be something like 
```
func internalServerError(w http.ResponseWriter) {
	http.Error(w, errors.New("something went wrong") http.StatusInternalServerError)
}
```
The reason being I don't want to propagate internal error messages to the user as its a user facing endpoint. so the middleware will propagate the error to user ?

Reply:
It's okay to propagate the internal error to the user for now. When we add the middleware, it will all work out. There are a few tricks one can use here, let's discuss when we add middleware (hopefully next week)

Reply:
ok